// be-fluent-school-backend/src/services/storage.js
import { drizzle } from "drizzle-orm/neon-http";
import { neon } from "@neondatabase/serverless";
import { eq, and, between, or, sql as sqlFunc, desc, asc } from "drizzle-orm";
import {
  users, profiles, teacherAvailability, classes, classReschedules, forumPosts, forumLikes, feedbacks,
  emailTemplates, emailSequences, emailSequenceSteps, marketingCampaigns,
  userEvents, leadScoring, leadScoringRules, abTests, abTestVariants,
  emailCampaignExecutions, marketingWorkflows, marketingAnalytics, announcements, paymentSettings, siteSettings,
  materials, meetLinks, messages, supportTickets, lessonProgress,
  activityProgress, exams
} from "../../shared/schema.js";

if (!process.env.DATABASE_URL) {
  throw new Error("DATABASE_URL is not defined");
}

const sql = neon(process.env.DATABASE_URL);
const db = drizzle(sql);


export class DatabaseStorage {
  // User methods
  async getUser(id) {
    try {
      const result = await db.select({
        id: users.id,
        username: users.username,
        password: users.password,
        must_change_password: users.must_change_password,
        password_reset_token: users.password_reset_token,
        password_reset_expires: users.password_reset_expires,
        created_at: users.created_at,
      }).from(users).where(eq(users.id, id)).limit(1);
      if (result.length > 0) {
        return result[0];
      }
      return undefined;
    } catch (error) {
      console.error('Error getting user by id:', {
        id,
        error: error.message,
        stack: error.stack,
        timestamp: new Date().toISOString()
      });
      throw error;
    }
  }

  async getUserByUsername(username) {
    try {
      console.log("DatabaseStorage.getUserByUsername called with:", username);
      const result = await db.select({
        id: users.id,
        username: users.username,
        password: users.password,
        must_change_password: users.must_change_password,
        password_reset_token: users.password_reset_token,
        password_reset_expires: users.password_reset_expires,
        created_at: users.created_at,
      }).from(users).where(eq(users.username, username)).limit(1);
      console.log("Query result count:", result.length);
      if (result.length > 0) {
        console.log("User found:", result[0].username);
        return result[0];
      } else {
        console.log("No user found with username:", username);
      }
      return undefined;
    } catch (error) {
      console.error('Error getting user by username:', {
        username,
        error: error.message,
        stack: error.stack,
        timestamp: new Date().toISOString()
      });
      throw error;
    }
  }

  async createUser(insertUser) {
    const result = await db.insert(users).values(insertUser).returning();
    return result[0];
  }

  async updateUser(id, user) {
    const result = await db.update(users).set(user).where(eq(users.id, id)).returning();
    return result[0];
  }

  async getUserByResetToken(token) {
    const result = await db
      .select()
      .from(users)
      .where(and(
        eq(users.password_reset_token, token),
        sqlFunc`${users.password_reset_expires} > now()`
      ))
      .limit(1);
    return result[0];
  }

  async setPasswordResetToken(userId, token, expires) {
    const result = await db
      .update(users)
      .set({
        password_reset_token: token,
        password_reset_expires: expires
      })
      .where(eq(users.id, userId))
      .returning();
    return result[0];
  }

  // Profile methods
  async getProfile(id) {
    const result = await db.select().from(profiles).where(eq(profiles.id, id)).limit(1);
    return result[0];
  }

  async getProfileByUserId(userId) {
    try {
      console.log("DatabaseStorage.getProfileByUserId called with:", userId);
      const result = await db.select().from(profiles).where(eq(profiles.user_id, userId)).limit(1);
      console.log("Profile query result count:", result.length);
      if (result.length > 0) {
        console.log("Profile found for user:", result[0].full_name, "role:", result[0].role);
      } else {
        console.log("No profile found for user ID:", userId);
      }
      return result[0];
    } catch (error) {
      console.error('Error getting profile by user ID:', {
        userId,
        error: error.message,
        stack: error.stack,
        timestamp: new Date().toISOString()
      });
      throw error;
    }
  }

  async createProfile(profile) {
    const result = await db.insert(profiles).values(profile).returning();
    return result[0];
  }

  async updateProfile(id, profile) {
    const result = await db.update(profiles).set(profile).where(eq(profiles.id, id)).returning();
    return result[0];
  }

  async getAllProfiles() {
    const result = await db.select().from(profiles)
      .where(eq(profiles.is_active, true))
      .orderBy(profiles.created_at);
    return result;
  }

  async getAllTeachers() {
    const result = await db.select().from(profiles)
      .where(and(eq(profiles.role, 'teacher'), eq(profiles.is_active, true)));
    return result;
  }

  // Student management methods
  async getAllStudents() {
    const result = await db
      .select()
      .from(profiles)
      .where(eq(profiles.role, 'student'))
      .orderBy(profiles.full_name);
    return result;
  }

  async getActiveStudents() {
    const result = await db
      .select()
      .from(profiles)
      .where(and(eq(profiles.role, 'student'), eq(profiles.is_active, true)))
      .orderBy(profiles.full_name);
    return result;
  }

  async getInactiveStudents() {
    const result = await db
      .select()
      .from(profiles)
      .where(and(eq(profiles.role, 'student'), eq(profiles.is_active, false)))
      .orderBy(profiles.full_name);
    return result;
  }

  async searchStudents(searchTerm) {
    const searchPattern = `%${searchTerm.toLowerCase()}%`;
    const result = await db
      .select()
      .from(profiles)
      .where(
        and(
          eq(profiles.role, 'student'),
          or(
            sqlFunc`LOWER(${profiles.full_name}) LIKE ${searchPattern}`,
            sqlFunc`LOWER(${profiles.email}) LIKE ${searchPattern}`,
            sqlFunc`LOWER(${profiles.cpf}) LIKE ${searchPattern}`,
            sqlFunc`LOWER(${profiles.phone}) LIKE ${searchPattern}`
          )
        )
      )
      .orderBy(profiles.full_name);
    return result;
  }

  async getStudentStats() {
    const result = await db
      .select({
        is_active: profiles.is_active,
        count: sqlFunc`count(*)`
      })
      .from(profiles)
      .where(eq(profiles.role, 'student'))
      .groupBy(profiles.is_active);

    // Get overdue payments count
    const overdueResult = await db
      .select({
        count: sqlFunc`count(*)`
      })
      .from(profiles)
      .where(
        and(
          eq(profiles.role, 'student'),
          eq(profiles.current_payment_status, 'overdue')
        )
      );

    const stats = { active: 0, inactive: 0, total: 0, overdue_payments: 0 };

    for (const row of result) {
      const count = Number(row.count);
      if (row.is_active) {
        stats.active = count;
      } else {
        stats.inactive = count;
      }
      stats.total += count;
    }

    stats.overdue_payments = Number(overdueResult[0]?.count || 0);

    return stats;
  }

  async deleteStudent(id) {
    const result = await db.delete(profiles).where(eq(profiles.id, id));
    return result.rowCount > 0;
  }

  async deleteUserByProfileId(profileId) {
    const profile = await this.getProfile(profileId);
    if (!profile) return false;

    // Atomic delete: profile first (removes FK dep), then user.
    await sql.transaction([
      sql`DELETE FROM profiles WHERE id = ${profileId}`,
      sql`DELETE FROM users WHERE id = ${profile.user_id}`,
    ]);

    return true;
  }

  // Teacher availability methods
  async getTeacherAvailability(teacherId) {
    const result = await db
      .select()
      .from(teacherAvailability)
      .where(eq(teacherAvailability.teacher_id, teacherId))
      .orderBy(teacherAvailability.day_of_week, teacherAvailability.start_time);
    return result;
  }

  async getTeacherAvailabilityById(id) {
    const result = await db
      .select()
      .from(teacherAvailability)
      .where(eq(teacherAvailability.id, id))
      .limit(1);
    return result[0];
  }

  async createTeacherAvailability(availability) {
    const result = await db.insert(teacherAvailability).values(availability).returning();
    return result[0];
  }

  async updateTeacherAvailability(id, availability) {
    const result = await db
      .update(teacherAvailability)
      .set({
        ...availability,
        updated_at: new Date()
      })
      .where(eq(teacherAvailability.id, id))
      .returning();
    return result[0];
  }

  async deleteTeacherAvailability(id) {
    const result = await db.delete(teacherAvailability).where(eq(teacherAvailability.id, id));
    return result.rowCount > 0;
  }

  // Payment Settings methods
  async getPaymentSettings() {
    const result = await db.select().from(paymentSettings).limit(1);
    if (result.length > 0) {
      return {
        activeProvider: 'asaas',
        asaasApiToken: result[0].asaas_api_token || '',
        asaasSandbox: result[0].asaas_sandbox
      };
    }
    return {
      activeProvider: 'asaas',
      asaasApiToken: '',
      asaasSandbox: true
    };
  }

  async savePaymentSettings(data, updatedBy) {
    const existing = await db.select().from(paymentSettings).limit(1);

    const settingsData = {
      active_provider: 'asaas',
      asaas_api_token: data.asaasApiToken,
      asaas_sandbox: data.asaasSandbox,
      updated_at: new Date(),
      updated_by: updatedBy
    };

    if (existing.length > 0) {
      const result = await db.update(paymentSettings)
        .set(settingsData)
        .where(eq(paymentSettings.id, existing[0].id))
        .returning();
      return result[0];
    } else {
      const result = await db.insert(paymentSettings).values(settingsData).returning();
      return result[0];
    }
  }

  // Site Settings methods (for branding, logos, banners)
  async getSiteSettings() {
    const result = await db.select().from(siteSettings).orderBy(siteSettings.category, siteSettings.setting_key);
    return result;
  }

  async getSiteSettingsByCategory(category) {
    const result = await db.select().from(siteSettings).where(eq(siteSettings.category, category));
    return result;
  }

  async getSiteSetting(key) {
    const result = await db.select().from(siteSettings).where(eq(siteSettings.setting_key, key)).limit(1);
    return result[0];
  }

  async setSiteSetting(key, value, type, category, label, description, updatedBy) {
    const existing = await db.select().from(siteSettings).where(eq(siteSettings.setting_key, key)).limit(1);
    
    if (existing.length > 0) {
      const result = await db.update(siteSettings)
        .set({
          setting_value: value,
          setting_type: type || existing[0].setting_type,
          category: category || existing[0].category,
          label: label || existing[0].label,
          description: description || existing[0].description,
          updated_by: updatedBy,
          updated_at: new Date()
        })
        .where(eq(siteSettings.setting_key, key))
        .returning();
      return result[0];
    } else {
      const result = await db.insert(siteSettings).values({
        setting_key: key,
        setting_value: value,
        setting_type: type || 'text',
        category: category || 'general',
        label: label || key,
        description: description,
        updated_by: updatedBy
      }).returning();
      return result[0];
    }
  }

  async deleteSiteSetting(key) {
    const result = await db.delete(siteSettings).where(eq(siteSettings.setting_key, key));
    return result.rowCount > 0;
  }

  async initializeDefaultSiteSettings(updatedBy) {
    const defaults = [
      // Branding
      { key: 'site_name', value: 'Be Fluent School', type: 'text', category: 'branding', label: 'Nome do Site', description: 'Nome exibido no site' },
      { key: 'site_logo', value: '', type: 'image', category: 'branding', label: 'Logo Principal', description: 'Logo exibido no cabeçalho' },
      { key: 'site_favicon', value: '', type: 'image', category: 'branding', label: 'Favicon', description: 'Ícone exibido na aba do navegador' },
      { key: 'primary_color', value: '#E59313', type: 'color', category: 'branding', label: 'Cor Principal', description: 'Cor primária do site' },
      { key: 'secondary_color', value: '#2B2B2B', type: 'color', category: 'branding', label: 'Cor Secundária', description: 'Cor secundária do site' },
      // Hero Section
      { key: 'hero_title', value: 'Domine o inglês que desbloqueia o próximo nível da sua carreira', type: 'text', category: 'hero', label: 'Título Principal', description: 'Título da seção hero' },
      { key: 'hero_subtitle', value: 'Construa liberdade comunicativa em inglês para reuniões, apresentações e oportunidades internacionais.', type: 'text', category: 'hero', label: 'Subtítulo', description: 'Subtítulo da seção hero' },
      { key: 'hero_cta_text', value: 'Agendar PowerTalk Grátis', type: 'text', category: 'hero', label: 'Texto do Botão CTA', description: 'Texto do botão de ação' },
      { key: 'hero_cta_link', value: 'https://wa.me/5598985332458', type: 'text', category: 'hero', label: 'Link do Botão CTA', description: 'Link de destino do botão' },
      { key: 'hero_background_image', value: '', type: 'image', category: 'hero', label: 'Imagem de Fundo', description: 'Imagem de fundo da seção hero' },
      // Footer
      { key: 'footer_text', value: 'Be Fluent School - Performance English', type: 'text', category: 'footer', label: 'Texto do Rodapé', description: 'Texto exibido no rodapé' },
      { key: 'whatsapp_number', value: '5598985332458', type: 'text', category: 'footer', label: 'Número WhatsApp', description: 'Número do WhatsApp para contato' },
      { key: 'instagram_url', value: '', type: 'text', category: 'footer', label: 'Link do Instagram', description: 'URL do perfil no Instagram' },
      // General
      { key: 'monthly_price', value: '350', type: 'text', category: 'general', label: 'Preço Mensalidade', description: 'Valor da mensalidade em reais' },
    ];

    for (const setting of defaults) {
      const existing = await db.select().from(siteSettings).where(eq(siteSettings.setting_key, setting.key)).limit(1);
      if (existing.length === 0) {
        await db.insert(siteSettings).values({
          setting_key: setting.key,
          setting_value: setting.value,
          setting_type: setting.type,
          category: setting.category,
          label: setting.label,
          description: setting.description,
          updated_by: updatedBy
        });
      }
    }
  }

  // Announcements methods
  async getAnnouncements() {
    const result = await db
      .select()
      .from(announcements)
      .orderBy(desc(announcements.created_at));
    return result;
  }

  async getAnnouncement(id) {
    const result = await db
      .select()
      .from(announcements)
      .where(eq(announcements.id, id))
      .limit(1);
    return result[0];
  }

  async createAnnouncement(data) {
    const result = await db.insert(announcements).values(data).returning();
    return result[0];
  }

  async updateAnnouncement(id, data) {
    const result = await db
      .update(announcements)
      .set(data)
      .where(eq(announcements.id, id))
      .returning();
    return result[0];
  }

  async deleteAnnouncement(id) {
    const result = await db.delete(announcements).where(eq(announcements.id, id));
    return result.rowCount > 0;
  }

  // Classes methods
  async getClass(id) {
    const result = await db.select().from(classes).where(eq(classes.id, id)).limit(1);
    return result[0];
  }

  async getClassesByTeacher(teacherId, startDate, endDate) {
    let whereCondition = eq(classes.teacher_id, teacherId);

    if (startDate && endDate) {
      whereCondition = and(
        eq(classes.teacher_id, teacherId),
        between(classes.scheduled_at, startDate, endDate)
      );
    }

    const result = await db
      .select()
      .from(classes)
      .where(whereCondition)
      .orderBy(classes.scheduled_at);
    return result;
  }

  async getClassesByStudent(studentId, startDate, endDate) {
    let whereCondition = eq(classes.student_id, studentId);

    if (startDate && endDate) {
      whereCondition = and(
        eq(classes.student_id, studentId),
        between(classes.scheduled_at, startDate, endDate)
      );
    }

    const result = await db
      .select()
      .from(classes)
      .where(whereCondition)
      .orderBy(classes.scheduled_at);
    return result;
  }

  async createClass(classData) {
    // Check if the class time is within teacher availability
    const isWithinAvailability = await this.checkTeacherAvailability(
      classData.teacher_id,
      classData.scheduled_at,
      classData.duration_minutes || 60
    );

    if (!isWithinAvailability) {
      throw new Error('Class scheduled outside teacher availability');
    }

    // Check for conflicts before creating
    const hasConflict = await this.checkClassConflict(
      classData.teacher_id,
      classData.scheduled_at,
      classData.duration_minutes || 60
    );

    if (hasConflict) {
      throw new Error('Schedule conflict detected');
    }

    const result = await db.insert(classes).values(classData).returning();
    return result[0];
  }

  async updateClass(id, classData) {
    // Check for conflicts if scheduled_at or duration is being updated
    if (classData.scheduled_at || classData.duration_minutes) {
      const existingClass = await this.getClass(id);
      if (existingClass) {
        const teacherId = classData.teacher_id || existingClass.teacher_id;
        const scheduledAt = classData.scheduled_at || existingClass.scheduled_at;
        const duration = classData.duration_minutes || existingClass.duration_minutes;

        // Check if the new time is within teacher availability
        const isWithinAvailability = await this.checkTeacherAvailability(
          teacherId,
          scheduledAt,
          duration
        );

        if (!isWithinAvailability) {
          throw new Error('Class scheduled outside teacher availability');
        }

        const hasConflict = await this.checkClassConflict(
          teacherId,
          scheduledAt,
          duration,
          id // exclude current class from conflict check
        );

        if (hasConflict) {
          throw new Error('Schedule conflict detected');
        }
      }
    }

    const result = await db
      .update(classes)
      .set({
        ...classData,
        updated_at: new Date()
      })
      .where(eq(classes.id, id))
      .returning();
    return result[0];
  }

  async deleteClass(id) {
    const result = await db.delete(classes).where(eq(classes.id, id));
    return result.rowCount > 0;
  }

  async checkClassConflict(teacherId, scheduledAt, duration, excludeClassId) {
    const classEndTime = new Date(scheduledAt.getTime() + duration * 60 * 1000);

    let whereCondition = and(
      eq(classes.teacher_id, teacherId),
      or(
        // New class starts during existing class
        and(
          sqlFunc`${classes.scheduled_at} <= ${scheduledAt}`,
          sqlFunc`(${classes.scheduled_at} + INTERVAL '1 minute' * ${classes.duration_minutes}) > ${scheduledAt}`
        ),
        // New class ends during existing class
        and(
          sqlFunc`${classes.scheduled_at} < ${classEndTime}`,
          sqlFunc`(${classes.scheduled_at} + INTERVAL '1 minute' * ${classes.duration_minutes}) >= ${classEndTime}`
        ),
        // Existing class is within new class timeframe
        and(
          sqlFunc`${classes.scheduled_at} >= ${scheduledAt}`,
          sqlFunc`(${classes.scheduled_at} + INTERVAL '1 minute' * ${classes.duration_minutes}) <= ${classEndTime}`
        )
      )
    );

    if (excludeClassId) {
      whereCondition = and(
        eq(classes.teacher_id, teacherId),
        sqlFunc`${classes.id} != ${excludeClassId}`,
        or(
          and(
            sqlFunc`${classes.scheduled_at} <= ${scheduledAt}`,
            sqlFunc`(${classes.scheduled_at} + INTERVAL '1 minute' * ${classes.duration_minutes}) > ${scheduledAt}`
          ),
          and(
            sqlFunc`${classes.scheduled_at} < ${classEndTime}`,
            sqlFunc`(${classes.scheduled_at} + INTERVAL '1 minute' * ${classes.duration_minutes}) >= ${classEndTime}`
          ),
          and(
            sqlFunc`${classes.scheduled_at} >= ${scheduledAt}`,
            sqlFunc`(${classes.scheduled_at} + INTERVAL '1 minute' * ${classes.duration_minutes}) <= ${classEndTime}`
          )
        )
      );
    }

    const result = await db
      .select({ id: classes.id })
      .from(classes)
      .where(whereCondition)
      .limit(1);

    return result.length > 0;
  }

  async checkStudentClassConflict(
    studentId,
    scheduledAt,
    duration,
    excludeClassId
  ) {
    const classEndTime = new Date(scheduledAt.getTime() + duration * 60 * 1000);

    let whereCondition = and(
      eq(classes.student_id, studentId),
      eq(classes.status, 'scheduled'), // Only check scheduled classes
      or(
        // New class starts during existing class
        and(
          sqlFunc`${classes.scheduled_at} <= ${scheduledAt}`,
          sqlFunc`(${classes.scheduled_at} + INTERVAL '1 minute' * ${classes.duration_minutes}) > ${scheduledAt}`
        ),
        // New class ends during existing class
        and(
          sqlFunc`${classes.scheduled_at} < ${classEndTime}`,
          sqlFunc`(${classes.scheduled_at} + INTERVAL '1 minute' * ${classes.duration_minutes}) >= ${classEndTime}`
        ),
        // Existing class is within new class timeframe
        and(
          sqlFunc`${classes.scheduled_at} >= ${scheduledAt}`,
          sqlFunc`(${classes.scheduled_at} + INTERVAL '1 minute' * ${classes.duration_minutes}) <= ${classEndTime}`
        )
      )
    );

    if (excludeClassId) {
      whereCondition = and(
        eq(classes.student_id, studentId),
        eq(classes.status, 'scheduled'),
        sqlFunc`${classes.id} != ${excludeClassId}`,
        or(
          and(
            sqlFunc`${classes.scheduled_at} <= ${scheduledAt}`,
            sqlFunc`(${classes.scheduled_at} + INTERVAL '1 minute' * ${classes.duration_minutes}) > ${scheduledAt}`
          ),
          and(
            sqlFunc`${classes.scheduled_at} < ${classEndTime}`,
            sqlFunc`(${classes.scheduled_at} + INTERVAL '1 minute' * ${classes.duration_minutes}) >= ${classEndTime}`
          ),
          and(
            sqlFunc`${classes.scheduled_at} >= ${scheduledAt}`,
            sqlFunc`(${classes.scheduled_at} + INTERVAL '1 minute' * ${classes.duration_minutes}) <= ${classEndTime}`
          )
        )
      );
    }

    const result = await db
      .select({ id: classes.id })
      .from(classes)
      .where(whereCondition)
      .limit(1);

    return result.length > 0;
  }

  async checkTeacherAvailability(teacherId, scheduledAt, duration) {
    const dayOfWeek = scheduledAt.getUTCDay();
    const startTime = this.formatUTCTime(scheduledAt); // HH:MM:SS format
    const endTime = this.formatUTCTime(new Date(scheduledAt.getTime() + duration * 60 * 1000));

    // Get teacher availability for the day
    const availability = await db
      .select()
      .from(teacherAvailability)
      .where(
        and(
          eq(teacherAvailability.teacher_id, teacherId),
          eq(teacherAvailability.day_of_week, dayOfWeek),
          eq(teacherAvailability.is_available, true),
          sqlFunc`${teacherAvailability.start_time} <= ${startTime}`,
          sqlFunc`${teacherAvailability.end_time} >= ${endTime}`
        )
      );

    return availability.length > 0;
  }

  async getAvailableSlots(teacherId, date, slotDuration = 60) {
    const dayOfWeek = date.getUTCDay();

    // Get teacher availability for the day
    const availability = await db
      .select()
      .from(teacherAvailability)
      .where(
        and(
          eq(teacherAvailability.teacher_id, teacherId),
          eq(teacherAvailability.day_of_week, dayOfWeek),
          eq(teacherAvailability.is_available, true)
        )
      );

    if (availability.length === 0) {
      return [];
    }

    // Get existing classes for the day using UTC-safe date range
    const startOfDay = new Date(date);
    startOfDay.setUTCHours(0, 0, 0, 0);
    const endOfDay = new Date(date);
    endOfDay.setUTCHours(23, 59, 59, 999);

    const existingClasses = await db
      .select()
      .from(classes)
      .where(
        and(
          eq(classes.teacher_id, teacherId),
          between(classes.scheduled_at, startOfDay, endOfDay)
        )
      )
      .orderBy(classes.scheduled_at);

    // Generate discrete available time slots
    const slots = [];

    for (const avail of availability) {
      // Convert time strings to minutes for easier calculation
      const startMinutes = this.timeStringToMinutes(avail.start_time);
      const endMinutes = this.timeStringToMinutes(avail.end_time);

      // Generate slots for this availability window
      for (let currentMinutes = startMinutes; currentMinutes + slotDuration <= endMinutes; currentMinutes += slotDuration) {
        const slotStart = this.minutesToTimeString(currentMinutes);
        const slotEnd = this.minutesToTimeString(currentMinutes + slotDuration);

        // Check if this slot conflicts with any existing class
        let hasConflict = false;
        for (const existingClass of existingClasses) {
          const classStart = this.formatUTCTime(existingClass.scheduled_at, false); // HH:MM format
          const classEndTime = new Date(existingClass.scheduled_at.getTime() + existingClass.duration_minutes * 60 * 1000);
          const classEnd = this.formatUTCTime(classEndTime, false);

          if (
            (slotStart >= classStart && slotStart < classEnd) ||
            (slotEnd > classStart && slotEnd <= classEnd) ||
            (slotStart = classEnd)
          ) {
            hasConflict = true;
            break;
          }
        }

        if (!hasConflict) {
          slots.push(slotStart);
        }
      }
    }

    return slots.sort();
  }

  // Helper methods for time calculations
  timeStringToMinutes(timeString) {
    const [hours, minutes] = timeString.split(':').map(Number);
    return hours * 60 + minutes;
  }

  minutesToTimeString(minutes) {
    const hours = Math.floor(minutes / 60);
    const mins = minutes % 60;
    return `${hours.toString().padStart(2, '0')}:${mins.toString().padStart(2, '0')}`;
  }

  // Helper method to format UTC time consistently
  formatUTCTime(date, includeSeconds = true) {
    const hours = date.getUTCHours().toString().padStart(2, '0');
    const minutes = date.getUTCMinutes().toString().padStart(2, '0');
    if (includeSeconds) {
      const seconds = date.getUTCSeconds().toString().padStart(2, '0');
      return `${hours}:${minutes}:${seconds}`;
    }
    return `${hours}:${minutes}`;
  }

  // Class reschedule methods
  async getClassReschedule(id) {
    const result = await db.select().from(classReschedules).where(eq(classReschedules.id, id)).limit(1);
    return result[0];
  }

  async getClassReschedules(classId) {
    const result = await db
      .select()
      .from(classReschedules)
      .where(eq(classReschedules.class_id, classId))
      .orderBy(classReschedules.created_at);
    return result;
  }

  async createClassReschedule(reschedule) {
    // Get the class details to validate the reschedule request
    const classData = await this.getClass(reschedule.class_id);
    if (!classData) {
      throw new Error('Class not found');
    }

    // Check if the new time is within teacher availability
    const isWithinAvailability = await this.checkTeacherAvailability(
      classData.teacher_id,
      reschedule.new_scheduled_at,
      classData.duration_minutes
    );

    if (!isWithinAvailability) {
      throw new Error('Reschedule request outside teacher availability');
    }

    // Check for conflicts with the new time
    const hasConflict = await this.checkClassConflict(
      classData.teacher_id,
      reschedule.new_scheduled_at,
      classData.duration_minutes,
      classData.id
    );

    if (hasConflict) {
      throw new Error('Reschedule request has schedule conflict');
    }

    // Update the class status to rescheduled
    await this.updateClass(classData.id, {
      status: 'rescheduled'
    });

    const result = await db.insert(classReschedules).values(reschedule).returning();
    return result[0];
  }

  async updateClassReschedule(id, reschedule) {
    // If approving a reschedule, check for conflicts and update the class
    if (reschedule.status === 'approved') {
      const existingReschedule = await this.getClassReschedule(id);
      if (existingReschedule) {
        const classData = await this.getClass(existingReschedule.class_id);
        if (classData) {
          // Check if the new time is within teacher availability
          const isWithinAvailability = await this.checkTeacherAvailability(
            classData.teacher_id,
            existingReschedule.new_scheduled_at,
            classData.duration_minutes
          );

          if (!isWithinAvailability) {
            throw new Error('Reschedule request outside teacher availability');
          }

          // Check for conflicts with the new time
          const hasConflict = await this.checkClassConflict(
            classData.teacher_id,
            existingReschedule.new_scheduled_at,
            classData.duration_minutes,
            classData.id
          );

          if (hasConflict) {
            throw new Error('Reschedule request has schedule conflict');
          }

          // Update the class with the new scheduled time
          await this.updateClass(classData.id, {
            scheduled_at: existingReschedule.new_scheduled_at,
            status: 'scheduled' // Reset status back to scheduled
          });
        }
      }
    }

    const result = await db
      .update(classReschedules)
      .set({
        ...reschedule,
        updated_at: new Date()
      })
      .where(eq(classReschedules.id, id))
      .returning();
    return result[0];
  }

  async getPendingReschedules(teacherId) {
    let whereCondition = eq(classReschedules.status, 'pending');

    if (teacherId) {
      whereCondition = and(
        eq(classReschedules.status, 'pending'),
        eq(classes.teacher_id, teacherId)
      );
    }

    const result = await db
      .select({
        reschedule: classReschedules,
        class: classes,
        student: profiles
      })
      .from(classReschedules)
      .innerJoin(classes, eq(classReschedules.class_id, classes.id))
      .innerJoin(profiles, eq(classes.student_id, profiles.id))
      .where(whereCondition)
      .orderBy(classReschedules.created_at);

    return result.map(r => r.reschedule);
  }

  async getStudentReschedules(studentId) {
    const result = await db
      .select({
        reschedule: classReschedules
      })
      .from(classReschedules)
      .innerJoin(classes, eq(classReschedules.class_id, classes.id))
      .where(eq(classes.student_id, studentId))
      .orderBy(sqlFunc`${classReschedules.created_at} DESC`);

    return result.map(r => r.reschedule);
  }

  // Forum methods
  async getForumPosts(authorId, category) {
    // Build where conditions
    let whereConditions = eq(forumPosts.parent_id, sqlFunc`NULL`); // Only top-level posts

    if (authorId) {
      whereConditions = and(whereConditions, eq(forumPosts.author_id, authorId));
    }

    const result = await db
      .select({
        id: forumPosts.id,
        title: forumPosts.title,
        content: forumPosts.content,
        author_id: forumPosts.author_id,
        class_id: forumPosts.class_id,
        parent_id: forumPosts.parent_id,
        is_question: forumPosts.is_question,
        is_answered: forumPosts.is_answered,
        likes_count: forumPosts.likes_count,
        created_at: forumPosts.created_at,
        updated_at: forumPosts.updated_at,
        author: profiles
      })
      .from(forumPosts)
      .innerJoin(profiles, eq(forumPosts.author_id, profiles.id))
      .where(whereConditions)
      .orderBy(sqlFunc`${forumPosts.created_at} DESC`);

    return result;
  }

  async getForumPost(id) {
    const result = await db
      .select({
        id: forumPosts.id,
        title: forumPosts.title,
        content: forumPosts.content,
        author_id: forumPosts.author_id,
        class_id: forumPosts.class_id,
        parent_id: forumPosts.parent_id,
        is_question: forumPosts.is_question,
        is_answered: forumPosts.is_answered,
        likes_count: forumPosts.likes_count,
        created_at: forumPosts.created_at,
        updated_at: forumPosts.updated_at,
        author: profiles
      })
      .from(forumPosts)
      .innerJoin(profiles, eq(forumPosts.author_id, profiles.id))
      .where(eq(forumPosts.id, id))
      .limit(1);

    return result[0];
  }

  async createForumPost(post) {
    const result = await db.insert(forumPosts).values(post).returning();
    return result[0];
  }

  async updateForumPost(id, post) {
    const result = await db.update(forumPosts).set(post).where(eq(forumPosts.id, id)).returning();
    return result[0];
  }

  async deleteForumPost(id) {
    const result = await db.delete(forumPosts).where(eq(forumPosts.id, id));
    return result.rowCount > 0;
  }

  async getForumReplies(parentId) {
    const result = await db
      .select({
        id: forumPosts.id,
        title: forumPosts.title,
        content: forumPosts.content,
        author_id: forumPosts.author_id,
        class_id: forumPosts.class_id,
        parent_id: forumPosts.parent_id,
        is_question: forumPosts.is_question,
        is_answered: forumPosts.is_answered,
        likes_count: forumPosts.likes_count,
        created_at: forumPosts.created_at,
        updated_at: forumPosts.updated_at,
        author: profiles
      })
      .from(forumPosts)
      .innerJoin(profiles, eq(forumPosts.author_id, profiles.id))
      .where(eq(forumPosts.parent_id, parentId))
      .orderBy(sqlFunc`${forumPosts.created_at} ASC`);

    return result;
  }

  // Forum likes methods
  async getUserLike(postId, userId) {
    const result = await db
      .select()
      .from(forumLikes)
      .where(and(eq(forumLikes.post_id, postId), eq(forumLikes.user_id, userId)))
      .limit(1);

    return result[0];
  }

  async addLike(postId, userId) {
    // Check if like already exists
    const existingLike = await this.getUserLike(postId, userId);
    if (existingLike) {
      return existingLike;
    }

    // Add the like
    const result = await db.insert(forumLikes).values({
      post_id: postId,
      user_id: userId
    }).returning();

    // Update the likes count on the post
    await db
      .update(forumPosts)
      .set({
        likes_count: sqlFunc`${forumPosts.likes_count} + 1`
      })
      .where(eq(forumPosts.id, postId));

    return result[0];
  }

  async removeLike(postId, userId) {
    // Check if like exists
    const existingLike = await this.getUserLike(postId, userId);
    if (!existingLike) {
      return false;
    }

    // Remove the like
    const result = await db
      .delete(forumLikes)
      .where(and(eq(forumLikes.post_id, postId), eq(forumLikes.user_id, userId)));

    if (result.rowCount > 0) {
      // Update the likes count on the post
      await db
        .update(forumPosts)
        .set({
          likes_count: sqlFunc`GREATEST(${forumPosts.likes_count} - 1, 0)`
        })
        .where(eq(forumPosts.id, postId));

      return true;
    }

    return false;
  }

  async getPostLikesCount(postId) {
    const result = await db
      .select({ likes_count: forumPosts.likes_count })
      .from(forumPosts)
      .where(eq(forumPosts.id, postId))
      .limit(1);

    return result[0]?.likes_count || 0;
  }

  // Feedback methods
  async getFeedback(id) {
    const feedback = await db.select().from(feedbacks).where(eq(feedbacks.id, id)).limit(1);
    if (feedback.length === 0) return undefined;

    const feedbackData = feedback[0];

    // Get fromUser
    const fromUser = await db.select().from(profiles).where(eq(profiles.id, feedbackData.from_user_id)).limit(1);
    if (fromUser.length === 0) return undefined;

    // Get toUser if exists
    let toUser = undefined;
    if (feedbackData.to_user_id) {
      const toUserResult = await db.select().from(profiles).where(eq(profiles.id, feedbackData.to_user_id)).limit(1);
      toUser = toUserResult[0];
    }

    return {
      ...feedbackData,
      fromUser: fromUser[0],
      toUser
    };
  }

  async getFeedbacks(userId, feedbackType) {
    let whereCondition = sqlFunc`1=1`;

    if (userId) {
      whereCondition = or(
        eq(feedbacks.from_user_id, userId),
        eq(feedbacks.to_user_id, userId)
      );
    }

    if (feedbackType) {
      whereCondition = feedbackType && userId
        ? and(whereCondition, eq(feedbacks.feedback_type, feedbackType))
        : eq(feedbacks.feedback_type, feedbackType);
    }

    const feedbacksData = await db
      .select()
      .from(feedbacks)
      .where(whereCondition)
      .orderBy(feedbacks.created_at);

    const result = [];
    for (const feedback of feedbacksData) {
      // Get fromUser
      const fromUserResult = await db.select().from(profiles).where(eq(profiles.id, feedback.from_user_id)).limit(1);
      if (fromUserResult.length === 0) continue;

      // Get toUser if exists
      let toUser = undefined;
      if (feedback.to_user_id) {
        const toUserResult = await db.select().from(profiles).where(eq(profiles.id, feedback.to_user_id)).limit(1);
        toUser = toUserResult[0];
      }

      result.push({
        ...feedback,
        fromUser: fromUserResult[0],
        toUser
      });
    }

    return result;
  }

  async getFeedbacksByUser(userId) {
    const feedbacksData = await db
      .select()
      .from(feedbacks)
      .where(eq(feedbacks.from_user_id, userId))
      .orderBy(feedbacks.created_at);

    const result = [];
    for (const feedback of feedbacksData) {
      // Get fromUser
      const fromUserResult = await db.select().from(profiles).where(eq(profiles.id, feedback.from_user_id)).limit(1);
      if (fromUserResult.length === 0) continue;

      // Get toUser if exists
      let toUser = undefined;
      if (feedback.to_user_id) {
        const toUserResult = await db.select().from(profiles).where(eq(profiles.id, feedback.to_user_id)).limit(1);
        toUser = toUserResult[0];
      }

      result.push({
        ...feedback,
        fromUser: fromUserResult[0],
        toUser
      });
    }

    return result;
  }

  async getFeedbacksForUser(userId) {
    const feedbacksData = await db
      .select()
      .from(feedbacks)
      .where(eq(feedbacks.to_user_id, userId))
      .orderBy(feedbacks.created_at);

    const result = [];
    for (const feedback of feedbacksData) {
      // Get fromUser
      const fromUserResult = await db.select().from(profiles).where(eq(profiles.id, feedback.from_user_id)).limit(1);
      if (fromUserResult.length === 0) continue;

      // Get toUser if exists
      let toUser = undefined;
      if (feedback.to_user_id) {
        const toUserResult = await db.select().from(profiles).where(eq(profiles.id, feedback.to_user_id)).limit(1);
        toUser = toUserResult[0];
      }

      result.push({
        ...feedback,
        fromUser: fromUserResult[0],
        toUser
      });
    }

    return result;
  }

  async createFeedback(feedback) {
    const result = await db.insert(feedbacks).values(feedback).returning();
    return result[0];
  }

  async updateFeedback(id, feedback) {
    const result = await db
      .update(feedbacks)
      .set({
        ...feedback,
        updated_at: new Date()
      })
      .where(eq(feedbacks.id, id))
      .returning();
    return result[0];
  }

  async respondToFeedback(id, response, respondedBy) {
    const result = await db
      .update(feedbacks)
      .set({
        response,
        responded_by: respondedBy,
        responded_at: new Date(),
        status: 'responded',
        updated_at: new Date()
      })
      .where(eq(feedbacks.id, id))
      .returning();
    return result[0];
  }

  async deleteFeedback(id) {
    const result = await db.delete(feedbacks).where(eq(feedbacks.id, id));
    return result.rowCount > 0;
  }

  async getFeedbackStats() {
    const totalResult = await db
      .select({ count: sqlFunc`count(*)` })
      .from(feedbacks);

    const byTypeResult = await db
      .select({
        feedback_type: feedbacks.feedback_type,
        count: sqlFunc`count(*)`
      })
      .from(feedbacks)
      .groupBy(feedbacks.feedback_type);

    const avgRatingResult = await db
      .select({ avg: sqlFunc`avg(rating)` })
      .from(feedbacks);

    const total = Number(totalResult[0]?.count || 0);
    const byType = {};

    for (const row of byTypeResult) {
      byType[row.feedback_type] = Number(row.count);
    }

    const avgRating = Number(avgRatingResult[0]?.avg || 0);

    return { total, by_type: byType, avg_rating: avgRating };
  }

  // Reports methods implementation - generating real data from database
  async getRevenueData(timeRange) {
    // Calculate months based on timeRange
    const monthsToFetch = timeRange === '1month' ? 1 : timeRange === '3months' ? 3 : timeRange === '6months' ? 6 : 12;

    const monthNames = ['Jan', 'Fev', 'Mar', 'Abr', 'Mai', 'Jun', 'Jul', 'Ago', 'Set', 'Out', 'Nov', 'Dez'];
    const revenueData = [];

    // Generate last N months of data
    for (let i = monthsToFetch - 1; i >= 0; i--) {
      const date = new Date();
      date.setMonth(date.getMonth() - i);
      const month = monthNames[date.getMonth()];

      // Get students count for this month (students created during this month)
      const startOfMonth = new Date(date.getFullYear(), date.getMonth(), 1);
      const endOfMonth = new Date(date.getFullYear(), date.getMonth() + 1, 0, 23, 59, 59, 999);

      const studentsResult = await db
        .select({ count: sqlFunc`count(*)` })
        .from(profiles)
        .where(
          and(
            eq(profiles.role, 'student'),
            between(profiles.created_at, startOfMonth, endOfMonth)
          )
        );

      const newStudents = Number(studentsResult[0]?.count || 0);

      // Get total active students up to this month
      const totalStudentsResult = await db
        .select({ count: sqlFunc`count(*)` })
        .from(profiles)
        .where(
          and(
            eq(profiles.role, 'student'),
            eq(profiles.is_active, true),
            sqlFunc`${profiles.created_at} <= ${endOfMonth}`
          )
        );

      const totalStudents = Number(totalStudentsResult[0]?.count || 0);

      // Calculate estimated revenue (students * average monthly fee)
      const avgMonthlyFee = 200; // Default monthly fee - could be calculated from actual fees
      const revenue = totalStudents * avgMonthlyFee;

      // Calculate retention rate (percentage of active students)
      const retentionRate = totalStudents > 0 ? Math.min(95, 80 + Math.random() * 15) : 0; // Simple estimation

      revenueData.push({
        month,
        revenue,
        students: totalStudents,
        retention: Math.round(retentionRate)
      });
    }

    return revenueData;
  }

  async getStudentProgressData(timeRange) {
    const monthsToFetch = timeRange === '1month' ? 1 : timeRange === '3months' ? 3 : timeRange === '6months' ? 6 : 12;
    const monthNames = ['Jan', 'Fev', 'Mar', 'Abr', 'Mai', 'Jun', 'Jul', 'Ago', 'Set', 'Out', 'Nov', 'Dez'];
    const progressData = [];

    for (let i = monthsToFetch - 1; i >= 0; i--) {
      const date = new Date();
      date.setMonth(date.getMonth() - i);
      const month = monthNames[date.getMonth()];

      const endOfMonth = new Date(date.getFullYear(), date.getMonth() + 1, 0, 23, 59, 59, 999);

      // Count students by level up to this month
      const levelCounts = await db
        .select({
          student_level: profiles.student_level,
          count: sqlFunc`count(*)`
        })
        .from(profiles)
        .where(
          and(
            eq(profiles.role, 'student'),
            eq(profiles.is_active, true),
            sqlFunc`${profiles.created_at} <= ${endOfMonth}`
          )
        )
        .groupBy(profiles.student_level);

      const levelStats = {
        beginners: 0,
        intermediate: 0,
        advanced: 0
      };

      for (const level of levelCounts) {
        const count = Number(level.count);
        if (level.student_level === 'beginner') {
          levelStats.beginners = count;
        } else if (level.student_level === 'intermediate') {
          levelStats.intermediate = count;
        } else if (level.student_level === 'advanced') {
          levelStats.advanced = count;
        }
      }

      progressData.push({
        month,
        ...levelStats
      });
    }

    return progressData;
  }

  async getTeacherPerformanceData() {
    const teachers = await db.select().from(profiles).where(eq(profiles.role, 'teacher'));
    const performanceData = [];

    for (const teacher of teachers) {
      // Count students assigned to this teacher (via classes)
      const studentCountResult = await db
        .select({
          count: sqlFunc`count(distinct ${classes.student_id})`
        })
        .from(classes)
        .where(eq(classes.teacher_id, teacher.id));

      const studentCount = Number(studentCountResult[0]?.count || 0);

      // Count classes this month for hours calculation
      const currentMonth = new Date();
      const startOfMonth = new Date(currentMonth.getFullYear(), currentMonth.getMonth(), 1);
      const endOfMonth = new Date(currentMonth.getFullYear(), currentMonth.getMonth() + 1, 0, 23, 59, 59, 999);

      const classesThisMonth = await db
        .select({
          total_duration: sqlFunc`sum(${classes.duration_minutes})`
        })
        .from(classes)
        .where(
          and(
            eq(classes.teacher_id, teacher.id),
            eq(classes.status, 'completed'),
            between(classes.scheduled_at, startOfMonth, endOfMonth)
          )
        );

      const totalMinutes = Number(classesThisMonth[0]?.total_duration || 0);
      const totalHours = Math.round(totalMinutes / 60);

      // Get average rating from feedbacks (fallback to demo rating if feedbacks table doesn't exist)
      let avgRating = 0;
      try {
        const ratingResult = await db
          .select({
            avg_rating: sqlFunc`avg(${feedbacks.rating})`
          })
          .from(feedbacks)
          .where(eq(feedbacks.to_user_id, teacher.id));

        avgRating = Number(ratingResult[0]?.avg_rating || 0);
      } catch (error) {
        // If feedbacks table doesn't exist or query fails, keep rating at 0.
        console.warn('Feedbacks rating unavailable for teacher:', teacher.full_name, error?.message || error);
        avgRating = 0;
      }

      // Calculate completion rate (estimated based on completed vs scheduled classes)
      const completedClassesResult = await db
        .select({ count: sqlFunc`count(*)` })
        .from(classes)
        .where(
          and(
            eq(classes.teacher_id, teacher.id),
            eq(classes.status, 'completed')
          )
        );

      const totalClassesResult = await db
        .select({ count: sqlFunc`count(*)` })
        .from(classes)
        .where(eq(classes.teacher_id, teacher.id));

      const completedClasses = Number(completedClassesResult[0]?.count || 0);
      const totalClasses = Number(totalClassesResult[0]?.count || 0);
      const completionRate = totalClasses > 0 ? Math.round((completedClasses / totalClasses) * 100) : 0;

      performanceData.push({
        name: teacher.full_name.split(' ')[0], // First name
        full_name: teacher.full_name,
        rating: Math.round(avgRating * 10) / 10, // Round to 1 decimal
        students: studentCount,
        completion: completionRate,
        hours: totalHours
      });
    }

    return performanceData;
  }

  async getCourseCompletionData() {
    // Get completion stats by student level
    const levelStats = await db
      .select({
        student_level: profiles.student_level,
        count: sqlFunc`count(*)`
      })
      .from(profiles)
      .where(
        and(
          eq(profiles.role, 'student'),
          eq(profiles.is_active, true)
        )
      )
      .groupBy(profiles.student_level);

    const totalStudents = levelStats.reduce((sum, level) => sum + Number(level.count), 0);

    const courseData = [
      { name: 'Básico', value: 0, color: '#10B981' },
      { name: 'Intermediário', value: 0, color: '#F59E0B' },
      { name: 'Avançado', value: 0, color: '#EF4444' },
      { name: 'Conversação', value: 0, color: '#8B5CF6' }
    ];

    if (totalStudents === 0) {
      return courseData; // Return empty data if no students
    }

    for (const level of levelStats) {
      const percentage = Math.round((Number(level.count) / totalStudents) * 100);

      if (level.student_level === 'beginner') {
        courseData[0].value = percentage;
      } else if (level.student_level === 'intermediate') {
        courseData[1].value = percentage;
      } else if (level.student_level === 'advanced') {
        courseData[2].value = percentage;
      }
    }

    // Conversação is estimated as a percentage of intermediate + advanced
    const intermediateAdvanced = courseData[1].value + courseData[2].value;
    courseData[3].value = Math.round(intermediateAdvanced * 0.6); // 60% of intermediate/advanced take conversation

    return courseData;
  }

  async getTodayClassesCount() {
    const today = new Date();
    const startOfDay = new Date(today);
    startOfDay.setHours(0, 0, 0, 0);
    const endOfDay = new Date(today);
    endOfDay.setHours(23, 59, 59, 999);

    const result = await db
      .select({ count: sqlFunc`count(*)` })
      .from(classes)
      .where(
        and(
          between(classes.scheduled_at, startOfDay, endOfDay),
          eq(classes.status, 'scheduled')
        )
      );

    return Number(result[0]?.count || 0);
  }

  // ===========================================
  // MARKETING AUTOMATION METHODS
  // ===========================================

  // Email Templates
  async getEmailTemplate(id) {
    const result = await db.select().from(emailTemplates).where(eq(emailTemplates.id, id)).limit(1);
    return result[0];
  }

  async getEmailTemplatesByType(type) {
    const result = await db.select().from(emailTemplates)
      .where(eq(emailTemplates.template_type, type))
      .orderBy(desc(emailTemplates.created_at));
    return result;
  }

  async getActiveEmailTemplates() {
    const result = await db.select().from(emailTemplates)
      .where(eq(emailTemplates.is_active, true))
      .orderBy(desc(emailTemplates.created_at));
    return result;
  }

  async createEmailTemplate(template) {
    const result = await db.insert(emailTemplates).values(template).returning();
    return result[0];
  }

  async updateEmailTemplate(id, template) {
    const result = await db.update(emailTemplates)
      .set({ ...template, updated_at: new Date() })
      .where(eq(emailTemplates.id, id))
      .returning();
    return result[0];
  }

  async deleteEmailTemplate(id) {
    const result = await db.delete(emailTemplates).where(eq(emailTemplates.id, id)).returning();
    return result.length > 0;
  }

  // Email Sequences
  async getEmailSequence(id) {
    const result = await db.select().from(emailSequences).where(eq(emailSequences.id, id)).limit(1);
    return result[0];
  }

  async getEmailSequenceByType(type) {
    const result = await db.select().from(emailSequences)
      .where(and(
        eq(emailSequences.sequence_type, type),
        eq(emailSequences.is_active, true)
      ))
      .limit(1);
    return result[0];
  }

  async getActiveEmailSequences() {
    const result = await db.select().from(emailSequences)
      .where(eq(emailSequences.is_active, true))
      .orderBy(desc(emailSequences.created_at));
    return result;
  }

  async createEmailSequence(sequence) {
    const result = await db.insert(emailSequences).values(sequence).returning();
    return result[0];
  }

  async updateEmailSequence(id, sequence) {
    const result = await db.update(emailSequences)
      .set({ ...sequence, updated_at: new Date() })
      .where(eq(emailSequences.id, id))
      .returning();
    return result[0];
  }

  async deleteEmailSequence(id) {
    const result = await db.delete(emailSequences).where(eq(emailSequences.id, id)).returning();
    return result.length > 0;
  }

  // Email Sequence Steps
  async getEmailSequenceStep(id) {
    const result = await db.select().from(emailSequenceSteps).where(eq(emailSequenceSteps.id, id)).limit(1);
    return result[0];
  }

  async getEmailSequenceSteps(sequenceId) {
    const result = await db.select().from(emailSequenceSteps)
      .where(and(
        eq(emailSequenceSteps.sequence_id, sequenceId),
        eq(emailSequenceSteps.is_active, true)
      ))
      .orderBy(asc(emailSequenceSteps.step_order));
    return result;
  }

  async createEmailSequenceStep(step) {
    const result = await db.insert(emailSequenceSteps).values(step).returning();
    return result[0];
  }

  async updateEmailSequenceStep(id, step) {
    const result = await db.update(emailSequenceSteps)
      .set(step)
      .where(eq(emailSequenceSteps.id, id))
      .returning();
    return result[0];
  }

  async deleteEmailSequenceStep(id) {
    const result = await db.delete(emailSequenceSteps).where(eq(emailSequenceSteps.id, id)).returning();
    return result.length > 0;
  }

  // Marketing Campaigns
  async getMarketingCampaign(id) {
    const result = await db.select().from(marketingCampaigns).where(eq(marketingCampaigns.id, id)).limit(1);
    return result[0];
  }

  async getMarketingCampaigns(status, type) {
    const conditions = [];
    if (status) conditions.push(eq(marketingCampaigns.status, status));
    if (type) conditions.push(eq(marketingCampaigns.campaign_type, type));

    const baseQuery = db.select().from(marketingCampaigns);

    const result = conditions.length > 0
      ? await baseQuery.where(and(...conditions)).orderBy(desc(marketingCampaigns.created_at))
      : await baseQuery.orderBy(desc(marketingCampaigns.created_at));

    return result;
  }

  async createMarketingCampaign(campaign) {
    const result = await db.insert(marketingCampaigns).values(campaign).returning();
    return result[0];
  }

  async updateMarketingCampaign(id, campaign) {
    const result = await db.update(marketingCampaigns)
      .set({ ...campaign, updated_at: new Date() })
      .where(eq(marketingCampaigns.id, id))
      .returning();
    return result[0];
  }

  async deleteMarketingCampaign(id) {
    const result = await db.delete(marketingCampaigns).where(eq(marketingCampaigns.id, id)).returning();
    return result.length > 0;
  }

  // User Events
  async getUserEvent(id) {
    const result = await db.select().from(userEvents).where(eq(userEvents.id, id)).limit(1);
    return result[0];
  }

  async getUserEvents(userId, limit = 50) {
    const result = await db.select().from(userEvents)
      .where(eq(userEvents.user_id, userId))
      .orderBy(desc(userEvents.timestamp))
      .limit(limit);
    return result;
  }

  async getUserEventsByType(userId, eventType, limit = 50) {
    const result = await db.select().from(userEvents)
      .where(and(
        eq(userEvents.user_id, userId),
        eq(userEvents.event_type, eventType)
      ))
      .orderBy(desc(userEvents.timestamp))
      .limit(limit);
    return result;
  }

  async createUserEvent(event) {
    const result = await db.insert(userEvents).values(event).returning();
    return result[0];
  }

  async getEventsSummary(userId, startDate, endDate) {
    const baseQuery = db.select({
      event_type: userEvents.event_type,
      count: sqlFunc`count(*)::int`,
    }).from(userEvents);

    const conditions = [eq(userEvents.user_id, userId)];
    if (startDate && endDate) {
      conditions.push(between(userEvents.timestamp, startDate, endDate));
    }

    const result = await baseQuery
      .where(and(...conditions))
      .groupBy(userEvents.event_type);

    return result.reduce((acc, row) => {
      acc[row.event_type] = row.count;
      return acc;
    }, {});
  }

  // Lead Scoring
  async getLeadScoring(id) {
    const result = await db.select().from(leadScoring).where(eq(leadScoring.id, id)).limit(1);
    return result[0];
  }

  async getLeadScoringByUserId(userId) {
    const result = await db.select().from(leadScoring).where(eq(leadScoring.user_id, userId)).limit(1);
    return result[0];
  }

  async getLeadsByTier(tier) {
    const result = await db.select().from(leadScoring)
      .where(eq(leadScoring.score_tier, tier))
      .orderBy(desc(leadScoring.total_score));
    return result;
  }

  async createLeadScoring(scoring) {
    const result = await db.insert(leadScoring).values(scoring).returning();
    return result[0];
  }

  async updateLeadScoring(userId, scoring) {
    const result = await db.update(leadScoring)
      .set({ ...scoring, updated_at: new Date() })
      .where(eq(leadScoring.user_id, userId))
      .returning();
    return result[0];
  }

  async getTopLeads(limit = 10) {
    const result = await db.select().from(leadScoring)
      .orderBy(desc(leadScoring.total_score))
      .limit(limit);
    return result;
  }

  // Lead Scoring Rules
  async getLeadScoringRule(id) {
    const result = await db.select().from(leadScoringRules).where(eq(leadScoringRules.id, id)).limit(1);
    return result[0];
  }

  async getLeadScoringRules() {
    const result = await db.select().from(leadScoringRules).orderBy(desc(leadScoringRules.created_at));
    return result;
  }

  async getActiveLeadScoringRules() {
    const result = await db.select().from(leadScoringRules)
      .where(eq(leadScoringRules.is_active, true))
      .orderBy(desc(leadScoringRules.created_at));
    return result;
  }

  async createLeadScoringRule(rule) {
    const result = await db.insert(leadScoringRules).values(rule).returning();
    return result[0];
  }

  async updateLeadScoringRule(id, rule) {
    const result = await db.update(leadScoringRules)
      .set({ ...rule, updated_at: new Date() })
      .where(eq(leadScoringRules.id, id))
      .returning();
    return result[0];
  }

  async deleteLeadScoringRule(id) {
    const result = await db.delete(leadScoringRules).where(eq(leadScoringRules.id, id)).returning();
    return result.length > 0;
  }

  // A/B Testing
  async getAbTest(id) {
    const result = await db.select().from(abTests).where(eq(abTests.id, id)).limit(1);
    return result[0];
  }

  async getAbTests(status) {
    const baseQuery = db.select().from(abTests);
    const result = status
      ? await baseQuery.where(eq(abTests.status, status)).orderBy(desc(abTests.created_at))
      : await baseQuery.orderBy(desc(abTests.created_at));
    return result;
  }

  async createAbTest(test) {
    const result = await db.insert(abTests).values(test).returning();
    return result[0];
  }

  async updateAbTest(id, test) {
    const result = await db.update(abTests)
      .set({ ...test, updated_at: new Date() })
      .where(eq(abTests.id, id))
      .returning();
    return result[0];
  }

  async deleteAbTest(id) {
    const result = await db.delete(abTests).where(eq(abTests.id, id)).returning();
    return result.length > 0;
  }

  // A/B Test Variants
  async getAbTestVariant(id) {
    const result = await db.select().from(abTestVariants).where(eq(abTestVariants.id, id)).limit(1);
    return result[0];
  }

  async getAbTestVariants(testId) {
    const result = await db.select().from(abTestVariants)
      .where(eq(abTestVariants.test_id, testId))
      .orderBy(asc(abTestVariants.variant_name));
    return result;
  }

  async createAbTestVariant(variant) {
    const result = await db.insert(abTestVariants).values(variant).returning();
    return result[0];
  }

  async updateAbTestVariant(id, variant) {
    const result = await db.update(abTestVariants)
      .set(variant)
      .where(eq(abTestVariants.id, id))
      .returning();
    return result[0];
  }

  async deleteAbTestVariant(id) {
    const result = await db.delete(abTestVariants).where(eq(abTestVariants.id, id)).returning();
    return result.length > 0;
  }

  // Email Campaign Executions
  async getEmailCampaignExecution(id) {
    const result = await db.select().from(emailCampaignExecutions).where(eq(emailCampaignExecutions.id, id)).limit(1);
    return result[0];
  }

  async getEmailCampaignExecutions(campaignId, userId) {
    const conditions = [];
    if (campaignId) conditions.push(eq(emailCampaignExecutions.campaign_id, campaignId));
    if (userId) conditions.push(eq(emailCampaignExecutions.user_id, userId));

    const baseQuery = db.select().from(emailCampaignExecutions);
    const result = conditions.length > 0
      ? await baseQuery.where(and(...conditions)).orderBy(desc(emailCampaignExecutions.sent_at))
      : await baseQuery.orderBy(desc(emailCampaignExecutions.sent_at));

    return result;
  }

  async createEmailCampaignExecution(execution) {
    const result = await db.insert(emailCampaignExecutions).values(execution).returning();
    return result[0];
  }

  async updateEmailExecutionTracking(id, trackingData) {
    const result = await db.update(emailCampaignExecutions)
      .set(trackingData)
      .where(eq(emailCampaignExecutions.id, id))
      .returning();
    return result[0];
  }

  // Marketing Workflows
  async getMarketingWorkflow(id) {
    const result = await db.select().from(marketingWorkflows).where(eq(marketingWorkflows.id, id)).limit(1);
    return result[0];
  }

  async getMarketingWorkflows(isActive) {
    const baseQuery = db.select().from(marketingWorkflows);
    const result = typeof isActive === 'boolean'
      ? await baseQuery.where(eq(marketingWorkflows.is_active, isActive)).orderBy(desc(marketingWorkflows.created_at))
      : await baseQuery.orderBy(desc(marketingWorkflows.created_at));
    return result;
  }

  async createMarketingWorkflow(workflow) {
    const result = await db.insert(marketingWorkflows).values(workflow).returning();
    return result[0];
  }

  async updateMarketingWorkflow(id, workflow) {
    const result = await db.update(marketingWorkflows)
      .set({ ...workflow, updated_at: new Date() })
      .where(eq(marketingWorkflows.id, id))
      .returning();
    return result[0];
  }

  async deleteMarketingWorkflow(id) {
    const result = await db.delete(marketingWorkflows).where(eq(marketingWorkflows.id, id)).returning();
    return result.length > 0;
  }

  // Marketing Analytics
  async getMarketingAnalytics(date, metricType) {
    const baseQuery = db.select().from(marketingAnalytics);

    const conditions = [eq(marketingAnalytics.date, date)];
    if (metricType) conditions.push(eq(marketingAnalytics.metric_type, metricType));

    const result = await baseQuery.where(and(...conditions));
    return result;
  }

  async createMarketingAnalytics(analytics) {
    const result = await db.insert(marketingAnalytics).values(analytics).returning();
    return result[0];
  }

  async getCampaignPerformance(campaignId) {
    const result = await db.select({
      total_sent: sqlFunc`count(*)::int`,
      total_delivered: sqlFunc`count(case when delivered_at is not null then 1 end)::int`,
      total_opened: sqlFunc`count(case when opened_at is not null then 1 end)::int`,
      total_clicked: sqlFunc`count(case when first_click_at is not null then 1 end)::int`,
      open_rate: sqlFunc`round((count(case when opened_at is not null then 1 end)::decimal / count(*)) * 100, 2)`,
      click_rate: sqlFunc`round((count(case when first_click_at is not null then 1 end)::decimal / count(*)) * 100, 2)`
    }).from(emailCampaignExecutions)
      .where(eq(emailCampaignExecutions.campaign_id, campaignId));

    return result[0];
  }

  async getEmailPerformance(templateId) {
    const result = await db.select({
      total_sent: sqlFunc`count(*)::int`,
      total_delivered: sqlFunc`count(case when delivered_at is not null then 1 end)::int`,
      total_opened: sqlFunc`count(case when opened_at is not null then 1 end)::int`,
      total_clicked: sqlFunc`count(case when first_click_at is not null then 1 end)::int`,
      open_rate: sqlFunc`round((count(case when opened_at is not null then 1 end)::decimal / count(*)) * 100, 2)`,
      click_rate: sqlFunc`round((count(case when first_click_at is not null then 1 end)::decimal / count(*)) * 100, 2)`
    }).from(emailCampaignExecutions)
      .where(eq(emailCampaignExecutions.template_id, templateId));

    return result[0];
  }

  async getConversionFunnel(startDate, endDate) {
    const result = await db.select({
      event_type: userEvents.event_type,
      count: sqlFunc`count(*)::int`
    }).from(userEvents)
      .where(between(userEvents.timestamp, startDate, endDate))
      .groupBy(userEvents.event_type);

    return result;
  }

  // ===========================================
  // MÉTODOS OBRIGATÓRIOS PARA MARKETING AUTOMATION
  // ===========================================

  async getScheduledCampaigns() {
    const now = new Date();
    const result = await db.select().from(marketingCampaigns)
      .where(and(
        eq(marketingCampaigns.status, 'scheduled'),
        sqlFunc`scheduled_at <= ${now}`
      ))
      .orderBy(asc(marketingCampaigns.scheduled_at));
    return result;
  }

  async getRunningAbTests() {
    const result = await db.select().from(abTests)
      .where(eq(abTests.status, 'running'))
      .orderBy(desc(abTests.created_at));
    return result;
  }

  async getCampaignsByAbTest(testId) {
    const result = await db.select().from(marketingCampaigns)
      .where(eq(marketingCampaigns.ab_test_id, testId));
    return result;
  }

  async incrementAbTestVariantMetric(variantId, metricType) {
    const updateData = {};
    switch (metricType) {
      case 'impression':
        updateData.impressions = sqlFunc`COALESCE(impressions, 0) + 1`;
        break;
      case 'sent':
        updateData.emails_sent = sqlFunc`COALESCE(emails_sent, 0) + 1`;
        break;
      case 'open':
        updateData.opens = sqlFunc`COALESCE(opens, 0) + 1`;
        break;
      case 'click':
        updateData.clicks = sqlFunc`COALESCE(clicks, 0) + 1`;
        break;
      case 'conversion':
        updateData.conversions = sqlFunc`COALESCE(conversions, 0) + 1`;
        break;
    }

    if (Object.keys(updateData).length > 0) {
      await db.update(abTestVariants)
        .set(updateData)
        .where(eq(abTestVariants.id, variantId));

      // Update conversion rate
      await db.update(abTestVariants)
        .set({
          conversion_rate: sqlFunc`CASE WHEN COALESCE(emails_sent, 0) > 0 THEN (COALESCE(conversions, 0)::decimal / COALESCE(emails_sent, 0)) * 100 ELSE 0 END`
        })
        .where(eq(abTestVariants.id, variantId));
    }
  }

  async getCampaignTargetAudience(campaignId) {
    // For now, get all active students as target audience
    // In real implementation, this would be based on campaign targeting criteria
    const result = await db.select().from(profiles)
      .where(and(
        eq(profiles.role, 'student'),
        eq(profiles.subscription_status, 'active')
      ))
      .orderBy(asc(profiles.created_at));
    return result;
  }

  async getEmailCampaignExecutionByMessageId(messageId) {
    const result = await db.select().from(emailCampaignExecutions)
      .where(eq(emailCampaignExecutions.sg_message_id, messageId))
      .limit(1);
    return result[0];
  }

  async getSuppressionListSize() {
    const result = await db.select({ count: sqlFunc`count(*)::int` })
      .from(emailSuppressions);
    return result[0]?.count || 0;
  }

  async getPendingEmailExecutions() {
    const now = new Date();
    const result = await db.select().from(emailCampaignExecutions)
      .where(and(
        eq(emailCampaignExecutions.delivery_status, 'pending'),
        or(
          eq(emailCampaignExecutions.scheduled_at, null),
          sqlFunc`scheduled_at <= ${now}`
        )
      ))
      .orderBy(asc(emailCampaignExecutions.created_at))
      .limit(100); // Process in batches
    return result;
  }

  async getInactiveLeads(inactiveThreshold) {
    const result = await db.select().from(leadScoring)
      .where(sqlFunc`last_activity < ${inactiveThreshold}`)
      .orderBy(desc(leadScoring.total_score));
    return result;
  }

  async stopABTest(testId, winnerVariantId) {
    const updateData = {
      status: 'stopped',
      ended_at: new Date(),
      updated_at: new Date()
    };

    if (winnerVariantId) {
      updateData.winner_variant_id = winnerVariantId;
    }

    const result = await db.update(abTests)
      .set(updateData)
      .where(eq(abTests.id, testId))
      .returning();
    return result[0];
  }

  // ===========================================
  // MÉTODOS PARA ANALYTICS E MÉTRICAS OBRIGATÓRIOS
  // ===========================================

  async getTotalCampaigns() {
    const result = await db.select({ count: sqlFunc`count(*)::int` })
      .from(marketingCampaigns);
    return result[0]?.count || 0;
  }

  async getTotalUsers() {
    const result = await db.select({ count: sqlFunc`count(*)::int` })
      .from(users);
    return result[0]?.count || 0;
  }

  async getTotalUserEvents() {
    const result = await db.select({ count: sqlFunc`count(*)::int` })
      .from(userEvents);
    return result[0]?.count || 0;
  }

  async getTotalLeads() {
    const result = await db.select({ count: sqlFunc`count(*)::int` })
      .from(leadScoring);
    return result[0]?.count || 0;
  }

  async getQualifiedLeads() {
    const result = await db.select({ count: sqlFunc`count(*)::int` })
      .from(leadScoring)
      .where(sqlFunc`total_score >= 50`); // Leads with score >= 50 are qualified
    return result[0]?.count || 0;
  }

  async getAverageLeadScore() {
    const result = await db.select({ avg: sqlFunc`avg(total_score)::int` })
      .from(leadScoring);
    return result[0]?.avg || 0;
  }

  async getCampaignMetrics() {
    const result = await db.select({
      campaign_id: emailCampaignExecutions.campaign_id,
      total_sent: sqlFunc`count(*)::int`,
      total_delivered: sqlFunc`count(case when delivered_at is not null then 1 end)::int`,
      total_opened: sqlFunc`count(case when opened_at is not null then 1 end)::int`,
      total_clicked: sqlFunc`count(case when first_click_at is not null then 1 end)::int`,
      open_rate: sqlFunc`round((count(case when opened_at is not null then 1 end)::decimal / count(*)) * 100, 2)`,
      click_rate: sqlFunc`round((count(case when first_click_at is not null then 1 end)::decimal / count(*)) * 100, 2)`
    }).from(emailCampaignExecutions)
      .groupBy(emailCampaignExecutions.campaign_id)
      .orderBy(desc(sqlFunc`count(*)`));

    return result;
  }

  // ─── Materials ───────────────────────────────────────────────────
  async getMaterials({ level, classId } = {}) {
    const conditions = [];
    if (level) conditions.push(eq(materials.level, level));
    if (classId) conditions.push(eq(materials.class_id, classId));
    const query = db.select().from(materials).orderBy(desc(materials.created_at));
    return conditions.length ? query.where(and(...conditions)) : query;
  }

  async getMaterialById(id) {
    const [material] = await db.select().from(materials).where(eq(materials.id, id)).limit(1);
    return material ?? null;
  }

  async createMaterial(data) {
    const [material] = await db.insert(materials).values(data).returning();
    return material;
  }

  async deleteMaterial(id) {
    const [deleted] = await db.delete(materials).where(eq(materials.id, id)).returning();
    return deleted ?? null;
  }

  async getRecentMarketingActivity() {
    const result = await db.select({
      type: sqlFunc`'campaign'`,
      id: marketingCampaigns.id,
      name: marketingCampaigns.name,
      status: marketingCampaigns.status,
      created_at: marketingCampaigns.created_at
    }).from(marketingCampaigns)
      .orderBy(desc(marketingCampaigns.created_at))
      .limit(10);

    return result;
  }

  // ─── Meet Links ───────────────────────────────────────────────────
  async getMeetLinks(createdBy) {
    const q = db.select().from(meetLinks).orderBy(desc(meetLinks.created_at));
    return createdBy ? q.where(eq(meetLinks.created_by, createdBy)) : q;
  }

  async createMeetLink(data) {
    const [link] = await db.insert(meetLinks).values(data).returning();
    return link;
  }

  async deleteMeetLink(id) {
    const [deleted] = await db.delete(meetLinks).where(eq(meetLinks.id, id)).returning();
    return deleted ?? null;
  }

  // ─── Messages ─────────────────────────────────────────────────────
  async getInboxMessages(userId) {
    return db.select().from(messages)
      .where(eq(messages.to_user_id, userId))
      .orderBy(desc(messages.created_at));
  }

  async getSentMessages(userId) {
    return db.select().from(messages)
      .where(eq(messages.from_user_id, userId))
      .orderBy(desc(messages.created_at));
  }

  async createMessage(data) {
    const [msg] = await db.insert(messages).values(data).returning();
    return msg;
  }

  async markMessageRead(id, userId) {
    const [updated] = await db.update(messages)
      .set({ is_read: true })
      .where(and(eq(messages.id, id), eq(messages.to_user_id, userId)))
      .returning();
    return updated ?? null;
  }

  async deleteMessage(id, userId) {
    const [deleted] = await db.delete(messages)
      .where(and(eq(messages.id, id), or(eq(messages.from_user_id, userId), eq(messages.to_user_id, userId))))
      .returning();
    return deleted ?? null;
  }

  // ─── Support Tickets ──────────────────────────────────────────────
  async getSupportTickets(userId, role) {
    if (role === 'admin') {
      return db.select().from(supportTickets).orderBy(desc(supportTickets.created_at));
    }
    return db.select().from(supportTickets)
      .where(eq(supportTickets.user_id, userId))
      .orderBy(desc(supportTickets.created_at));
  }

  async createSupportTicket(data) {
    const [ticket] = await db.insert(supportTickets).values(data).returning();
    return ticket;
  }

  async updateSupportTicketStatus(id, status) {
    const [updated] = await db.update(supportTickets)
      .set({ status, updated_at: new Date() })
      .where(eq(supportTickets.id, id))
      .returning();
    return updated ?? null;
  }

  // ─── Lesson Progress ──────────────────────────────────────────────
  async getLessonProgress(studentId) {
    return db.select().from(lessonProgress)
      .where(eq(lessonProgress.student_id, studentId));
  }

  async upsertLessonProgress(data) {
    const [result] = await db.insert(lessonProgress)
      .values(data)
      .onConflictDoUpdate({
        target: [lessonProgress.student_id, lessonProgress.lesson_id],
        set: {
          status: data.status,
          score: data.score,
          xp_earned: data.xp_earned,
          lesson_number: data.lesson_number,
          completed_at: data.completed_at,
          updated_at: new Date(),
        },
      })
      .returning();
    return result;
  }

  // ─── Progress Summary (calculated) ───────────────────────────────
  async getStudentProgressSummary(studentId) {
    const now = new Date();
    const weekAgo = new Date(now.getTime() - 7 * 24 * 60 * 60 * 1000);
    const monthAgo = new Date(now.getFullYear(), now.getMonth(), 1);

    const [allClasses, weekClasses, progressRows] = await Promise.all([
      db.select().from(classes).where(and(eq(classes.student_id, studentId), eq(classes.status, 'completed'))),
      db.select().from(classes).where(and(
        eq(classes.student_id, studentId),
        eq(classes.status, 'completed'),
        between(classes.scheduled_at, weekAgo, now)
      )),
      db.select().from(lessonProgress).where(eq(lessonProgress.student_id, studentId)),
    ]);

    const totalXP = progressRows.reduce((sum, r) => sum + (r.xp_earned || 0), 0) + allClasses.length * 100;
    const completedLessons = progressRows.filter(r => r.status === 'completed').length;

    const level = Math.max(1, Math.floor(totalXP / 500) + 1);
    const currentLevelXP = totalXP % 500;
    const nextLevelXP = 500;

    const studyMinutes = allClasses.reduce((sum, c) => sum + (c.duration_minutes || 60), 0);
    const weekStudyMinutes = weekClasses.reduce((sum, c) => sum + (c.duration_minutes || 60), 0);

    return {
      totalXP,
      level,
      currentLevelXP,
      nextLevelXP,
      totalCompletedClasses: allClasses.length,
      weekCompletedClasses: weekClasses.length,
      completedLessons,
      studyMinutes,
      weekStudyMinutes,
    };
  }

  // ─── Upcoming Classes (next N hours) ─────────────────────────────
  async getUpcomingClasses(hoursAhead = 48) {
    const now = new Date();
    const future = new Date(now.getTime() + hoursAhead * 60 * 60 * 1000);

    const rows = await db
      .select({
        id: classes.id,
        scheduled_at: classes.scheduled_at,
        status: classes.status,
        notes: classes.notes,
        duration_minutes: classes.duration_minutes,
        student_id: classes.student_id,
        teacher_id: classes.teacher_id,
      })
      .from(classes)
      .where(
        and(
          between(classes.scheduled_at, now, future),
          eq(classes.status, 'scheduled')
        )
      )
      .orderBy(asc(classes.scheduled_at))
      .limit(10);

    // Enrich with profile names
    const enriched = await Promise.all(rows.map(async (c) => {
      const [student, teacher] = await Promise.all([
        c.student_id ? db.select({ full_name: profiles.full_name }).from(profiles).where(eq(profiles.id, c.student_id)).limit(1) : [],
        c.teacher_id ? db.select({ full_name: profiles.full_name }).from(profiles).where(eq(profiles.id, c.teacher_id)).limit(1) : [],
      ]);
      return {
        ...c,
        student_name: student[0]?.full_name ?? null,
        teacher_name: teacher[0]?.full_name ?? null,
      };
    }));

    return enriched;
  }

  // ─── Admin Dashboard Consolidated Stats ──────────────────────────
  async getAdminDashboardStats() {
    const now = new Date();
    const startOfMonth = new Date(now.getFullYear(), now.getMonth(), 1);

    const [totalStudentsRes, activeTeachersRes, monthlyRevenueRes, pendingPaymentsRes, overduePaymentsRes, recentClassesRes] = await Promise.all([
      db.select({ count: sqlFunc`count(*)` }).from(profiles).where(and(eq(profiles.role, 'student'), eq(profiles.is_active, true))),
      db.select({ count: sqlFunc`count(*)` }).from(profiles).where(and(eq(profiles.role, 'teacher'), eq(profiles.is_active, true))),
      // Sum of monthly fees for active students
      db.select({ total: sqlFunc`coalesce(sum(monthly_fee), 0)` }).from(profiles).where(and(eq(profiles.role, 'student'), eq(profiles.is_active, true))),
      db.select({ count: sqlFunc`count(*)` }).from(profiles).where(and(eq(profiles.role, 'student'), eq(profiles.current_payment_status, 'pending'))),
      db.select({ count: sqlFunc`count(*)` }).from(profiles).where(and(eq(profiles.role, 'student'), eq(profiles.current_payment_status, 'overdue'))),
      // Recent classes this month
      db.select({ count: sqlFunc`count(*)` }).from(classes).where(and(
        between(classes.scheduled_at, startOfMonth, now),
        eq(classes.status, 'completed')
      )),
    ]);

    return {
      totalStudents: Number(totalStudentsRes[0]?.count ?? 0),
      activeTeachers: Number(activeTeachersRes[0]?.count ?? 0),
      monthlyRevenue: Number(monthlyRevenueRes[0]?.total ?? 0),
      pendingPayments: Number(pendingPaymentsRes[0]?.count ?? 0),
      overduePayments: Number(overduePaymentsRes[0]?.count ?? 0),
      classesThisMonth: Number(recentClassesRes[0]?.count ?? 0),
    };
  }

  // ─── Activity Progress ────────────────────────────────────────────
  async getActivityProgress(studentId, moduleId = null) {
    const conditions = [eq(activityProgress.student_id, studentId)];
    if (moduleId) conditions.push(eq(activityProgress.module_id, moduleId));
    return db.select().from(activityProgress).where(and(...conditions))
      .orderBy(asc(activityProgress.module_id), asc(activityProgress.activity_number));
  }

  async upsertActivityProgress(data) {
    const existing = await db.select().from(activityProgress)
      .where(and(
        eq(activityProgress.student_id, data.student_id),
        eq(activityProgress.module_id, data.module_id),
        eq(activityProgress.activity_number, data.activity_number),
      )).limit(1);

    if (existing.length > 0) {
      const updateData = { ...data, updated_at: new Date() };
      if (data.status === 'completed' && !existing[0].completed_at) {
        updateData.completed_at = new Date();
      }
      if (data.status === 'in_progress' && !existing[0].started_at) {
        updateData.started_at = new Date();
      }
      const [updated] = await db.update(activityProgress).set(updateData)
        .where(eq(activityProgress.id, existing[0].id)).returning();
      return updated;
    }

    const insertData = {
      ...data,
      started_at: data.status !== 'not_started' ? new Date() : null,
      completed_at: data.status === 'completed' ? new Date() : null,
    };
    const [inserted] = await db.insert(activityProgress).values(insertData).returning();
    return inserted;
  }

  async initializeStudentActivities(studentId, moduleId, startingActivity = 1) {
    // Create the first activity as in_progress, rest as not_started
    for (let i = 1; i <= 20; i++) {
      const status = i === startingActivity ? 'in_progress' : (i < startingActivity ? 'completed' : 'not_started');
      await this.upsertActivityProgress({ student_id: studentId, module_id: moduleId, activity_number: i, status, classes_used: 0 });
    }
  }

  // Check milestone and trigger exam notification if needed
  // Returns { triggered: bool, examType: string }
  async checkActivityMilestone(studentId, moduleId, completedActivityNumber, teacherId) {
    const EXAM_MILESTONES = [6, 11, 20];
    if (!EXAM_MILESTONES.includes(completedActivityNumber)) return { triggered: false };

    // Verify ALL activities up to this number are actually completed
    const rows = await db.select().from(activityProgress)
      .where(and(eq(activityProgress.student_id, studentId), eq(activityProgress.module_id, moduleId)));

    const completedSet = new Set(rows.filter(r => r.status === 'completed').map(r => r.activity_number));
    for (let i = 1; i <= completedActivityNumber; i++) {
      if (!completedSet.has(i)) return { triggered: false }; // gap found — not all done
    }

    // Check no exam already triggered for this milestone
    const existing = await db.select().from(exams).where(and(
      eq(exams.student_id, studentId),
      eq(exams.module_id, moduleId),
      eq(exams.triggered_by_activity, completedActivityNumber),
      eq(exams.exam_type, 'theoretical'),
    )).limit(1);
    if (existing.length > 0) return { triggered: false };

    // Create theoretical exam record
    await db.insert(exams).values({
      student_id: studentId,
      teacher_id: teacherId || null,
      module_id: moduleId,
      triggered_by_activity: completedActivityNumber,
      exam_type: 'theoretical',
      status: 'pending',
    });

    // Send notification message to student
    const student = await this.getProfile(studentId);
    if (student) {
      const milestoneLabel = completedActivityNumber === 20 ? 'módulo completo' : `atividade ${completedActivityNumber}`;
      await this.createMessage({
        from_user_id: teacherId,
        to_user_id: studentId,
        subject: `📝 Prova Teórica liberada — ${moduleId}`,
        content: `Parabéns! Você concluiu a ${milestoneLabel} do módulo ${moduleId}. Sua prova teórica foi liberada. Fique atento ao link do Google Forms que seu professor enviará. Tire suas dúvidas antes da prova! 🎯`,
        priority: 'high',
      });
    }

    return { triggered: true, examType: 'theoretical', activityNumber: completedActivityNumber };
  }

  // ─── Exams ────────────────────────────────────────────────────────
  async getStudentExams(studentId, moduleId = null) {
    const conditions = [eq(exams.student_id, studentId)];
    if (moduleId) conditions.push(eq(exams.module_id, moduleId));
    return db.select().from(exams).where(and(...conditions))
      .orderBy(desc(exams.triggered_at));
  }

  async getTeacherStudentExams(teacherId) {
    // Get all exams for students belonging to this teacher
    const teacherStudents = await db.select({ id: profiles.id })
      .from(profiles).where(and(eq(profiles.role, 'student'), eq(profiles.is_active, true)));
    // Filter by teacher_id in exam or by teacher relationship
    return db.select().from(exams).where(eq(exams.teacher_id, teacherId))
      .orderBy(desc(exams.triggered_at));
  }

  async updateExam(examId, data) {
    const updateData = { ...data, updated_at: new Date() };
    if (data.status === 'completed' && !updateData.completed_at) {
      updateData.completed_at = new Date();
    }
    const [updated] = await db.update(exams).set(updateData)
      .where(eq(exams.id, examId)).returning();

    // If theoretical exam completed, auto-create performatic exam
    if (data.status === 'completed' && updated?.exam_type === 'theoretical') {
      const existingPerformatic = await db.select().from(exams).where(and(
        eq(exams.student_id, updated.student_id),
        eq(exams.module_id, updated.module_id),
        eq(exams.triggered_by_activity, updated.triggered_by_activity),
        eq(exams.exam_type, 'performatic'),
      )).limit(1);

      if (existingPerformatic.length === 0) {
        await db.insert(exams).values({
          student_id: updated.student_id,
          teacher_id: updated.teacher_id,
          module_id: updated.module_id,
          triggered_by_activity: updated.triggered_by_activity,
          exam_type: 'performatic',
          status: 'pending',
        });

        // Notify student
        if (updated.teacher_id) {
          await this.createMessage({
            from_user_id: updated.teacher_id,
            to_user_id: updated.student_id,
            subject: `🎤 Análise Performática liberada — ${updated.module_id}`,
            content: `Sua prova teórica foi corrigida! Sua Análise Performática (prova oral) está liberada. Seu professor entrará em contato para agendar. Prepare-se bem! 💪`,
            priority: 'high',
          });
        }
      }
    }

    return updated;
  }

  async createExam(data) {
    const [inserted] = await db.insert(exams).values(data).returning();
    return inserted;
  }
}

export const storage = new DatabaseStorage();
